skema/skema-rs/mathml/src/expression.rs

```diff
@@ -14,42 +23,1094 @@ enum Atom {
+                                            parent_node_index,
+                                            op_copy[i + 1].to_string(),
+                                        );
+                                    } else {
+                                        graph.add_edge(
+                                            node_idx,
+                                            parent_node_index,
+                                            op[0].to_string(),
+                                        );
+                                    }
+                                } else {
+                                    graph.add_edge(
+                                        node_idx,
+                                        parent_node_index,
+                                        op_copy[i].to_string(),
+                                    );
+                                }
+                            }
+                            args[i].to_graph(graph);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub fn redundant_paren(str: &String) -> bool {
+    let str_len = get_str_len(str);
+    if str.chars().nth(0) != Some('(') || str.chars().nth(str_len - 1) != Some(')') {
+        return false;
+    }
+    let mut par_stack = VecDeque::new();
+    par_stack.push_back("left_par");
+    for i in 1..=str_len - 2 {
+        if str.chars().nth(i) == Some('(') {
+            par_stack.push_back("par");
+        } else if str.chars().nth(i) == Some(')') {
+            par_stack.pop_back();
+        }
+    }
+    if par_stack.len() > 0 {
+        if par_stack[0] == "left_par" {
+            return true;
+        }
+    }
+    return false;
+}
+
+pub fn all_multi_div(op: &mut Vec<Operator>) -> bool {
+    for o in 1..=op.len() - 1 {
+        if op[o] != Operator::Multiply && op[o] != Operator::Divide {
+            return false;
+        }
+    }
+    return true;
+}
+
+impl PreExp {
+    fn group_expr(&mut self) {
+        for arg in &mut self.args {
+            match arg {
+                Expr::Atom(_) => {}
+                Expr::Expression { .. } => {
+                    arg.group_expr();
+                }
+            }
+        }
+    }
+
+    fn get_names(&mut self) {
+        for mut arg in &mut self.args {
+            match &mut arg {
+                Expr::Atom(_) => {}
+                Expr::Expression { .. } => {
+                    arg.get_names();
+                }
+            }
+        }
+    }
+
+    fn to_graph(&mut self) -> MathMLGraph {
+        let mut g = MathMLGraph::new();
+        // let root_index = g.add_node("root".to_string());
+        for mut arg in &mut self.args {
+            match &mut arg {
+                Expr::Atom(_) => {}
+                Expr::Expression { .. } => {
+                    arg.to_graph(&mut g);
+                }
+            }
+        }
+        return g;
+    }
+}
+
+pub fn get_str_len(str: &str) -> usize {
```

Comment:
Why not simply use the `len` method (https://doc.rust-lang.org/std/primitive.str.html#method.len) instead of implementing your own function to calculate the length of a string slice?

Reply:
Calling the len() function directly will return the incorrect number of characters in a string. I've been finding and debugging this issue for a long time and had to write my own function.

Reply:
@ualiangzhang  Are you sure? Can you provide a minimum working example? 

I just tried changing `get_str_len(name)` to `name.len()` on line 252 and `get_str_len(str)` to `str.len()` on line 449 of `expression.rs`, and the tests are still passing.

Also, could you change the function arguments named `str` to `string` (since `str` is a Rust builtin, it would be good to avoid using it as a variable name - for one, it messes with my syntax highlighting :) )